mjpc/planners/cem/planner.cc

```diff
@@ -0,0 +1,527 @@
+
+  // improvement: compare nominal to winner
+  double best_return = trajectory[0].total_return;
+  improvement = mju_max(best_return - trajectory[winner].total_return, 0.0);
+
+  // stop timer
+  policy_update_compute_time = GetDuration(policy_update_start);
+}
+
+// compute trajectory using nominal policy
+void CEMPlanner::NominalTrajectory(int horizon, ThreadPool& pool) {
+  // set policy
+  auto nominal_policy = [&cp = candidate_policy[0]](
+                            double* action, const double* state, double time) {
+    cp.Action(action, state, time);
+  };
+
+  // rollout nominal policy
+  trajectory[0].Rollout(nominal_policy, task, model, data_[0].get(),
+                        state.data(), time, mocap.data(), userdata.data(),
+                        horizon);
+}
+
+// set action from policy
+void CEMPlanner::ActionFromPolicy(double* action, const double* state,
+                                       double time, bool use_previous) {
+  const std::shared_lock<std::shared_mutex> lock(mtx_);
+  if (use_previous) {
+    previous_policy.Action(action, state, time);
+  } else {
+    policy.Action(action, state, time);
+  }
+}
+
+// update policy via resampling
+void CEMPlanner::UpdateNominalPolicy(int horizon) {
+  // dimensions
+  int num_spline_points = candidate_policy[winner].num_spline_points;
+
+  // set time
+  double nominal_time = time;
+  double time_shift = mju_max(
+      (horizon - 1) * model->opt.timestep / (num_spline_points - 1), 1.0e-5);
+
+  // temp variables to help with averaging
+  std::vector<double> temp_avg(model->nu);  // holds avg for spline point t
+  // std::vector<double> temp_elite_actions(model->nu);
+  // std::array<std::array<double, nu>, n_elites> temp_elite_actions;
+  std::vector<std::vector<double>> temp_elite_actions;
+  temp_elite_actions.resize(n_elites, std::vector<double>(model->nu, 0.0));
+
+  // get spline points
+  for (int t = 0; t < num_spline_points; t++) {
+    times_scratch[t] = nominal_time;
+
+    // get the actions of the top n_elites policies and average them
+    // also update a variance parameter too
+    for (int i = 0; i < n_elites; i++) {
+      int index = trajectory_order[i];  // the (i+1)^th best trajectory [TODO] this segfaults!
+      candidate_policy[index].Action(DataAt(temp_elite_actions[i], 0), nullptr, nominal_time);
+      for (int k = 0; k < model->nu; k++) {
+        temp_avg[k] += temp_elite_actions[i][k] / num_spline_points;
+      }
+    }
+
+    // computing the sample variance of the control parameters
+    for (int k = 0; k < model->nu; k++) {
+      for (int i = 0; i < n_elites; i++) {
+        stdev[t * model->nu + k] += std::pow(temp_elite_actions[i][k] - temp_avg[k], 2) / (n_elites - 1);
+      }
+      stdev[t * model->nu + k] = std::sqrt(stdev[t * model->nu + k]);  // take the root
+    }
+
+    // assigning the averaged parameters to parameters_scratch
+    std::copy(temp_avg.begin(), temp_avg.end(), parameters_scratch.begin() + t * model->nu);
```

Comment:
this is the main meat of the CEM impl. For each spline point, we compute the sample mean and variance over the elite samples, where we've assumed that there is some good trajectory order already.